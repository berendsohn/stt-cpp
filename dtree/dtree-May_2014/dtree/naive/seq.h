// -*-C++-*-
// Copyright (c) 2012--2014 David Eisenstat <eisenstatdavid@gmail.com>
//     and Brown University
// Released under http://opensource.org/licenses/MIT
// May 2014 version

#ifndef DTREE_NAIVE_SEQ_H_
#define DTREE_NAIVE_SEQ_H_

#include "dtree/naive/common.h"

namespace naive {

template<typename Base>
class EndSeq : public Base {
 public:
  EndSeq() : solid_() {}
  EndSeq* solid(int i) const { return solid_[i]; }
  void set_solid(int i, EndSeq* solid_i) { solid_[i] = solid_i; }
  template<typename Forwarder>
  void CopyFrom(const EndSeq& node, Forwarder forward);
 private:
  EndSeq* solid_[2];
  DISALLOW_COPY_AND_ASSIGN(EndSeq);
};

template<typename Group, typename Base>
class WithValue : public Base {
 public:
  typedef Group Group_;
  typename Group::Type value() const { return value_; }
  void set_value(typename Group::Type value) { value_ = value; }
  void add_to_value(typename Group::Type delta) {
    set_value(Group::Plus(value(), delta));
  }
  void subtract_from_value(typename Group::Type delta) {
    set_value(Group::Minus(value(), delta));
  }
  template<typename Node>
  static typename Group::Type Value(Node* node);
  template<typename Node>
  static void SetValue(Node* node, typename Group::Type value);
  template<typename Node>
  static void AddToSeq(Node* node, typename Group::Type delta);
  template<typename Node>
  static void SubtractFromSeq(Node* node, typename Group::Type delta);

 protected:
  WithValue() : value_() {}

 private:
  typename Group::Type value_;
};

template<typename Semigroup, typename Base>
class WithAggr : public Base {
 public:
  typedef Semigroup Semigroup_;
  typename Semigroup::Type singleton_aggr() const {
    return Semigroup::AggrFromValue(this->value());
  }
  template<typename Node>
  static typename Semigroup::Type AggrSeq(Node* node);
  template<typename Node, typename Predicate>
  static Node* FindDirmostSeq(int dir, Node* node, const Predicate& predicate);
  template<typename Node, typename Predicate>
  static Node* FindLeftmostSeq(Node* node, const Predicate& predicate);
  template<typename Node, typename Predicate>
  static Node* FindRightmostSeq(Node* node, const Predicate& predicate);
 protected:
  WithAggr() {}
};

template<typename Base>
class WithReverseBy : public Base {
 private:
  typedef typename Base::Group_ Group;
 public:
  template<typename Node>
  static void ReverseSeq(Node* node);
 protected:
  WithReverseBy() {}
};

template<typename Base>
class WithReverse : public Base {
 public:
  template<typename Node>
  static void ReverseSeq(Node* node);
 protected:
  WithReverse() {}
};
}  // namespace naive
#endif  // DTREE_NAIVE_SEQ_H_
