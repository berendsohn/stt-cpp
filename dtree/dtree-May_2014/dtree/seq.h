// -*-C++-*-
// Copyright (c) 2012--2014 David Eisenstat <eisenstatdavid@gmail.com>
//     and Brown University
// Released under http://opensource.org/licenses/MIT
// May 2014 version

#ifndef DTREE_SEQ_H_
#define DTREE_SEQ_H_

#include "dtree/selector.h"

#ifdef NAIVE_DTREE
#include "dtree/naive/seq.h"
#else
#include "dtree/selfadjust/seq.h"
#endif
#endif  // DTREE_SEQ_H_
