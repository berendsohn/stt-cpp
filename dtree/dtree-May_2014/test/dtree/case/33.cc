#define DTREE_CONFIG EndSeq<WithValue<FreeGroup, WithValue<dtree::DpAdd<double>, WithAggr<dtree::Count<int>, dtree::Index_<int>, Begin > > > >
#include "test/dtree/seq.cc"
