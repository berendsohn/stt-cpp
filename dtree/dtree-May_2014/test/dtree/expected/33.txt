<<<
  9: vL10R14F0 v703,000,ead,000
  1: vL-6R-10F0 vbcf,000,958,000
  5: vL20R18F0 vd9e,000,325,000
  12: vL16R1F1 va70,000,084,000
  14: vL-2R-11F0 vbe1,000,9d6,000
  2: vL33R-12F0 v5c2,000,529,000
  4: vL-5R-2F0 vb6b,000,43d,000
  8: vL18R60F0 v2dd,000,c78,000
  7: vL21R16F1 v887,000,1b6,000
  10: vL10R-15F1 v4e7,000,5ce,000
  15: vL-6R1F0 v0c1,000,620,000
  11: vL9R54F1 v0b2,000,fd3,000
  3: vL3R-8F0 v523,000,029,000

  13: vL10R1F1 vf13,000,04d,000
  6: vL-4R-4F1 v921,000,133,000
>>>
<<<
      /9: a2 vL-10R-4F0 v703,d9e,ead,325
        \1: a1 vL-16R-24F0 vbcf,703,958,ead
    /5: a6 vL25R20F0 vd9e,b6b,325,43d
      \12: a3 vL-4R-17F1 va70,d9e,084,325
        \14: a2 vL-12R-18F1 vbe1,a70,9d6,084
          \2: a1 vL35R-1F0 v5c2,be1,529,9d6
  @4: a3 vL-5R-2F0 vb6b,000,43d,000
      /8: a2 vL75R8F1 v2dd,4e7,c78,5ce
        \7: a1 vL3R-44F1 v887,2dd,1b6,c78
    \10: a6 vL15R-13F1 v4e7,b6b,5ce,43d
      \15: a3 vL16R-16F1 v0c1,4e7,620,5ce
        \11: a2 vL15R53F1 v0b2,0c1,fd3,620
          \3: a1 vL-62R-6F1 v523,0b2,029,fd3

    /13: a1 vL5R14F0 vf13,921,04d,133
  @6: a12 vL-4R-4F1 v921,000,133,000
>>>
